### PR TITLE
Fix several integer related issues

### LINKED ISSUE #32508: Compiler doesn't complain about assigning some invalid values to builtin subtypes of `int`
**Description:**
The criteria for a value `v` to belong to a built-in subtype of `int` is given in [https://ballerina.io/spec/lang/2021R1/#built-in_subtypes](https://ballerina.io/spec/lang/2021R1/#built-in_subtypes). But some invalid values are being allowed to be assigned to some of the subtypes. Consider the following examples 

**Steps to reproduce:**
```ballerina

public function main() {
    // Signed16 -32768 ≤ v ≤ 32767
    //int:Signed16 x1 = 21474803711; // error
    int:Signed16 x2 = 21474803712; // no error
    int:Signed16 x3 = 9223372036854775807; // no error
    //int:Signed16 x4 = 9223372036854775808; // error
}

public function main() {
    // Signed8 -128 ≤ v ≤ 127
    //int:Signed8 x5 = -9223372036854775680; // error
    int:Signed8 x6 = -9223372036854775681; // no error
    int:Signed8 x7 = 9223372036854775807; // no error
    //int:Signed8 x8 = 9223372036854775808; // error
}

public function main() {
    // Unsigned16 0 ≤ v ≤ 65535
    //int:Unsigned16 x9 = -21474770944; // error
    int:Unsigned16 x10 = -21474770945; // no error
    int:Unsigned16 x11 = -21474836480; // no error
    //int:Unsigned16 x12 = -21474836481; // error
}

public function main() { 
    // Unsigned8 0 ≤ v ≤ 255
    //int:Unsigned8 x13 = -9223372036854775552; // error
    int:Unsigned8 x14 = -9223372036854775553; // no error
    int:Unsigned8 x15 = -9223372036854775808; // no error
    //int:Unsigned8 x16 = -9223372036854775809; // error
}
```
**Related Issues:**
https://github.com/ballerina-platform/ballerina-lang/issues/32574

**Affected Versions:**
slbeta2

### LINKED ISSUE #32574: Compiler crashes when a too large or too small number is assigned to a builtin subtype of `int`
**Description:**
Compiler crashes when a value > 9223372036854775807 or value < -9223372036854775808 is assigned to a variable of type Signed32, Signed16, Signed8, Unsigned32, Unsigned16, or Unsigned8

**Steps to reproduce:**
```ballerina
public function main() {
    int:Signed8 x = 9223372036854775808; // gives a compilation error, but also crashes
}
```
Error log:
```
SEVERE {b7a.log.crash} - class java.lang.String cannot be cast to class java.lang.Long (java.lang.String and java.lang.Long are in module java.base of loader 'bootstrap') 
java.lang.ClassCastException: class java.lang.String cannot be cast to class java.lang.Long (java.lang.String and java.lang.Long are in module java.base of loader 'bootstrap')
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.getIntLiteralType(TypeChecker.java:815)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.setLiteralValueAndGetType(TypeChecker.java:479)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.visit(TypeChecker.java:413)
	at org.wso2.ballerinalang.compiler.tree.expressions.BLangLiteral.accept(BLangLiteral.java:50)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkExpr(TypeChecker.java:376)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkExpr(TypeChecker.java:355)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:871)
	at org.wso2.ballerinalang.compiler.tree.BLangSimpleVariable.accept(BLangSimpleVariable.java:53)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3497)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3465)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeDef(SemanticAnalyzer.java:3457)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:1562)
	at org.wso2.ballerinalang.compiler.tree.statements.BLangSimpleVariableDef.accept(BLangSimpleVariableDef.java:46)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3497)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3465)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeStmt(SemanticAnalyzer.java:3461)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:470)
	at org.wso2.ballerinalang.compiler.tree.BLangBlockFunctionBody.accept(BLangBlockFunctionBody.java:57)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3497)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:440)
	at org.wso2.ballerinalang.compiler.tree.BLangFunction.accept(BLangFunction.java:73)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3497)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3465)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeDef(SemanticAnalyzer.java:3457)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:327)
	at org.wso2.ballerinalang.compiler.tree.BLangPackage.accept(BLangPackage.java:165)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyze(SemanticAnalyzer.java:292)
	at io.ballerina.projects.internal.CompilerPhaseRunner.typeCheck(CompilerPhaseRunner.java:204)
	at io.ballerina.projects.internal.CompilerPhaseRunner.performTypeCheckPhases(CompilerPhaseRunner.java:112)
	at io.ballerina.projects.ModuleContext.compileInternal(ModuleContext.java:383)
	at io.ballerina.projects.ModuleCompilationState$1.compile(ModuleCompilationState.java:45)
	at io.ballerina.projects.ModuleContext.compile(ModuleContext.java:329)
	at io.ballerina.projects.PackageCompilation.compileModulesInternal(PackageCompilation.java:174)
	at io.ballerina.projects.PackageCompilation.compileModules(PackageCompilation.java:166)
	at io.ballerina.projects.PackageCompilation.from(PackageCompilation.java:95)
	at io.ballerina.projects.PackageContext.getPackageCompilation(PackageContext.java:206)
	at io.ballerina.projects.Package.getCompilation(Package.java:135)
	at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:64)
	at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
	at io.ballerina.cli.cmd.RunCommand.execute(RunCommand.java:163)
	at java.base/java.util.Optional.ifPresent(Optional.java:183)
	at io.ballerina.cli.launcher.Main.main(Main.java:58)
```

The same issue is observed when the value is too small as well. 

```ballerina
public function main() {
    int:Signed16 x = -9223372036854775809;
}
```

**Affected Versions:**
slbeta2

### LINKED ISSUE #32577: Assigning `-9223372036854775808` to an `int` does not overflow
**Description:**
Consider the following line in [lang.int](https://ballerina.io/spec/lang/master/#lang.int) spec
```
# Minimum value of type `int`.
public const MIN_VALUE = -9223372036854775807 - 1; // -9223372036854775808 would overflow
```
Since ballerina doesn't have a concept of positives and negatives,  the positive part here `9223372036854775808` should overflow. 

**Steps to reproduce:**
```ballerina
public function main() {
    int a = -9223372036854775808; //This should overflow, but works ATM
}
```
**Affected Versions:**
slbeta2

### LINKED ISSUE #32686: Can not assign valid value to union of builtin subtypes of `int`
**Description:**
$title. Consider the following examples

**Steps to reproduce:**
```ballerina
import ballerina/io;

public function main() {
    int:Signed8|int:Unsigned32 a1 = 235262;
    io:println(a1); // prints 235262
    
    int:Signed8 x = -127;
    int:Signed8|int:Unsigned32 a2 = x;
    io:println(a2); // prints -127 

    int:Signed8|int:Unsigned32 a3 = -127; // incompatible types: expected 'int:Unsigned32', found 'int'
}
```
```ballerina
import ballerina/io;

type IntType int:Signed8|int:Unsigned32;

public function main() {
    IntType a1 = 235262;
    io:println(a1); // prints 235262
    
    int:Signed8 x = -127;
    IntType a2 = x;
    io:println(a2); // prints -127 

    IntType a3 = -127; // incompatible types: expected 'int:Unsigned32', found 'int'
}
```
**Affected Versions:**
slbeta2

### LINKED ISSUE #33689: Assignability of valid value to union of builtin subtypes of int depends on the order of definition
**Description:**
$title

**Steps to reproduce:**
```
import ballerina/io;

public function main() {
    int:Signed8|int:Unsigned32 a1 = 235262; // works
    int:Unsigned32|int:Signed8 a2 = 235262; // incompatible types: expected 'int:Signed8', found 'int'
}
```

**Related Issues (optional):**
#32686

### PR DESCRIPTION
**Pull request by suleka96**
Fixes #32577
Fixes  #32574
Fixes  #32508
Fixes #32686
Fixes #33689
Tests added #32884

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java [hunk 1]
```diff
@@ -5860,23 +5861,35 @@ private Object getIntegerLiteral(Node literal, String nodeValue, SyntaxKind sign
...
-                                        pos.lineRange().startLine().offset() - 1,
-                                        pos.lineRange().endLine().offset());
-                dlog.error(pos, code1, originalNodeValue);
-            } else {
-                dlog.error(pos, code2, originalNodeValue);
-            }
+            logParseLongError(literal, originalNodeValue, sign, code1, code2);
         }
         return originalNodeValue;
     }
```
Comment (chiranSachintha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java:5895]: This logic should move to `typechecker`. according to this(https://github.com/ballerina-platform/ballerina-spec/issues/943) `program 1` is a compile-time error and `program 2` should compile fine. 

Program 1
```ballerina
public function main() {
    int k = 999999999999999999999999999999;
}
```
Program 2
```ballerina
public function main() {
    float k = 999999999999999999999999999999;
}
```

**Comment by KRVPerera** [comment]
I think these error messages are misleading. We are saying `Integer` and it is too small. As soon as we identify the literal as an `Integer` it hints that this is a valid Integer, now the question is it is too small for what? . 

```
 Integer '-9223372036854775808' too small
 ```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeBuilder.java [hunk 2]
```diff
@@ -5841,35 +5842,28 @@ private DocumentationReferenceType stringToRefType(String refTypeName) {
...
     }
 
-    private Object parseLong(Node literal, String originalNodeValue,
-                             String processedNodeValue, int radix, SyntaxKind sign,
-                             DiagnosticCode code1, DiagnosticCode code2) {
+    private Object parseLong(String originalNodeValue, String processedNodeValue, int radix) {
+        // Since 9223372036854775808 is to overflow in Ballerina and -9223372036854775808 is a unary-expr where the
+        // expression is 9223372036854775808, -9223372036854775808 should also overflow even though it is
+        // within bounds for Java Long
+        if (originalNodeValue.equals(INT_MIN_OVERFLOW_VAL)) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeBuilder.java:5857]: As discussed offline, we shouldn't have to special case this here IMO. We should have modeled this as a unary expression (as opposed to how we model it as a literal atm).

For something like

```ballerina
int _ = -9223372036854775808;
```

the error should be for `9223372036854775808` and not the unary expression.

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeBuilder.java:5857]: Shall we check if doing this now is possible?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 3]
```diff
@@ -499,13 +499,17 @@ private BType getFiniteTypeMatchWithIntType(BLangLiteral literalExpr, BFiniteTyp
                 }
             }
         }
+        if (literalExpr.value instanceof String) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:502]: Can do this as the first check?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 4]
```diff
@@ -499,13 +499,17 @@ private BType getFiniteTypeMatchWithIntType(BLangLiteral literalExpr, BFiniteTyp
...
+            resultType = symTable.semanticError;
+            return resultType;
+        }
         return symTable.noType;
     }
 
     private BType getFiniteTypeMatchWithIntLiteral(BLangLiteral literalExpr, BFiniteType finiteType,
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:509]: The following should work, right?

```ballerina
type Foo 1|2|9223372036854775808;

public function main() {
    Foo _ = 9223372036854775808;
}
```

Currently results in an error.

```cmd
ERROR [x.bal:(4:13,4:32)] '9223372036854775808' is out of range
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeBuilder.java (hunk 2 above)
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeBuilder.java:5857]: Also related to https://github.com/ballerina-platform/ballerina-lang/issues/30020.

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeBuilder.java:5857]:
```ballerina
public function main() {
    int _ = -(9223372036854775808);
}
```

used to fail as expected with `ERROR [x.bal:(2:15,2:34)] Integer '9223372036854775808' too large`.

Crashes now.

```
[2021-12-07 16:04:01,243] SEVERE {b7a.log.crash} - class java.lang.String cannot be cast to class java.lang.Long (java.lang.String and java.lang.Long are in module java.base of loader 'bootstrap') 
java.lang.ClassCastException: class java.lang.String cannot be cast to class java.lang.Long (java.lang.String and java.lang.Long are in module java.base of loader 'bootstrap')
        at org.wso2.ballerinalang.compiler.bir.writer.BIRInstructionWriter.visit(BIRInstructionWriter.java:329)
        at org.wso2.ballerinalang.compiler.bir.model.BIRNonTerminator$ConstantLoad.accept(BIRNonTerminator.java:176)
        at org.wso2.ballerinalang.compiler.bir.writer.BIRInstructionWriter.lambda$visit$1(BIRInstructionWriter.java:137)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at org.wso2.ballerinalang.compiler.bir.writer.BIRInstructionWriter.visit(BIRInstructionWriter.java:131)
        at org.wso2.ballerinalang.compiler.bir.model.BIRNode$BIRBasicBlock.accept(BIRNode.java:443)
        at org.wso2.ballerinalang.compiler.bir.writer.BIRInstructionWriter.lambda$writeBBs$0(BIRInstructionWriter.java:87)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at org.wso2.ballerinalang.compiler.bir.writer.BIRInstructionWriter.writeBBs(BIRInstructionWriter.java:87)
        at org.wso2.ballerinalang.compiler.bir.writer.BIRBinaryWriter.writeFunction(BIRBinaryWriter.java:297)
        at org.wso2.ballerinalang.compiler.bir.writer.BIRBinaryWriter.lambda$writeFunctions$5(BIRBinaryWriter.java:197)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at org.wso2.ballerinalang.compiler.bir.writer.BIRBinaryWriter.writeFunctions(BIRBinaryWriter.java:197)
        at org.wso2.ballerinalang.compiler.bir.writer.BIRBinaryWriter.serialize(BIRBinaryWriter.java:86)
        at io.ballerina.projects.ModuleContext.cacheBIR(ModuleContext.java:490)
        at io.ballerina.projects.ModuleContext.generateCodeInternal(ModuleContext.java:460)
        at io.ballerina.projects.ModuleCompilationState$4.generatePlatformSpecificCode(ModuleCompilationState.java:132)
        at io.ballerina.projects.ModuleContext.generatePlatformSpecificCode(ModuleContext.java:389)
        at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:159)
        at io.ballerina.projects.JBallerinaBackend.<init>(JBallerinaBackend.java:137)
        at io.ballerina.projects.JBallerinaBackend.lambda$from$0(JBallerinaBackend.java:113)
        at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1133)
        at io.ballerina.projects.PackageCompilation.getCompilerBackend(PackageCompilation.java:182)
        at io.ballerina.projects.JBallerinaBackend.from(JBallerinaBackend.java:112)
        at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:90)
        at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
        at io.ballerina.cli.cmd.RunCommand.execute(RunCommand.java:203)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at io.ballerina.cli.launcher.Main.main(Main.java:51)
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeBuilder.java [hunk 5]
```diff
@@ -5240,19 +5235,25 @@ private BLangLiteral createSimpleLiteral(Node literal) {
...
+//    private BLangLiteral createSimpleLiteral(Node literal, boolean isFiniteType) {
+//        if (literal.kind() == SyntaxKind.UNARY_EXPRESSION) {
+//            UnaryExpressionNode unaryExpr = (UnaryExpressionNode) literal;
+//            BLangLiteral bLangLiteral =
+//                    createSimpleLiteral(unaryExpr.expression(), unaryExpr.unaryOperator().kind(), isFiniteType);
+//            bLangLiteral.pos = getPosition(unaryExpr); // setting the proper pos, else only the expr pos is set
+//            return bLangLiteral;
+//        }
+//        return createSimpleLiteral(literal, SyntaxKind.NONE, isFiniteType);
+//    }
```
Comment (rdulmina) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeBuilder.java:5247]: Let's remove unwanted code

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 6]
```diff
@@ -4547,10 +4580,79 @@ public void visit(BLangUnaryExpr unaryExpr) {
                 actualType = new BTypedescType(exprType, null);
             }
         } else {
-            //allow both addition and subtraction operators to get expected type as Decimal
-            boolean decimalAddNegate = expType.tag == TypeTags.DECIMAL &&
-                    (OperatorKind.ADD.equals(unaryExpr.operator) || OperatorKind.SUB.equals(unaryExpr.operator));
-            exprType = decimalAddNegate ? checkExpr(unaryExpr.expr, env, expType) : checkExpr(unaryExpr.expr, env);
+            //allow both addition and subtraction operators to get expected type
+            boolean isExpTypeAllowed = (OperatorKind.ADD.equals(unaryExpr.operator)
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:4584]: Why do we need to include `+`?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 7]
```diff
@@ -4547,10 +4580,79 @@ public void visit(BLangUnaryExpr unaryExpr) {
...
+                        || referredType.tag == TypeTags.DECIMAL) {
+                    newExpectedType = referredType;
+                } else if (referredType.tag == TypeTags.FINITE) {
+                    BFiniteType finiteType = (BFiniteType) referredType;
+                    for (BLangExpression value : finiteType.getValueSpace()) {
+                        int typeTag = value.getBType().tag;
+                        if (TypeTags.isIntegerTypeTag(typeTag) || typeTag == TypeTags.FLOAT
+                                || typeTag == TypeTags.DECIMAL) {
+                            basicNumericTypes.add(value.getBType());
+                        } else if (typeTag == TypeTags.JSON || typeTag == TypeTags.ANYDATA || typeTag == TypeTags.ANY) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:4605]: How can the type tag of a member of a finite type be any of these?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 8]
```diff
@@ -4547,10 +4580,79 @@ public void visit(BLangUnaryExpr unaryExpr) {
...
+                        int typeTag = value.tag;
+                        if (TypeTags.isIntegerTypeTag(typeTag) || typeTag == TypeTags.FLOAT
+                                || typeTag == TypeTags.DECIMAL) {
+                            basicNumericTypes.add(value);
+                        } else if (typeTag == TypeTags.JSON || typeTag == TypeTags.ANYDATA || typeTag == TypeTags.ANY) {
+                            basicNumericTypes.add(symTable.intType);
+                            basicNumericTypes.add(symTable.floatType);
+                            basicNumericTypes.add(symTable.decimalType);
+                            break;
+                        }
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:4633]: If we extract out the relevant numeric type identification logic to a separate method, we can recursively call that here.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 9]
```diff
@@ -4547,10 +4580,79 @@ public void visit(BLangUnaryExpr unaryExpr) {
...
+                    if (basicNumericTypes.isEmpty()) {
+                        dlog.error(unaryExpr.pos, DiagnosticErrorCode.UNARY_OP_INCOMPATIBLE_TYPES,
+                                unaryExpr.operator, unaryExpr.expr);
+                        resultType = symTable.semanticError;
+                    } else if (basicNumericTypes.size() == 1) {
+                        newExpectedType = basicNumericTypes.iterator().next();
+                    } else {
+                        newExpectedType = BUnionType.create(null, basicNumericTypes);
+                    }
+                } else if (referredType.tag == TypeTags.JSON || referredType.tag == TypeTags.ANYDATA ||
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:4644]: The following fails at the backend atm.

```ballerina
anydata _ = -1234;
```

```
error: negation is not supported for type: (int|float|decimal)
```

**Comment by MaryamZi** [comment]
@suleka96, please change the base to [stage-swan-lake](https://github.com/ballerina-platform/ballerina-lang/tree/stage-swan-lake).

**Comment by MaryamZi** [comment]
Is https://github.com/ballerina-platform/ballerina-lang/issues/32686 fixed? The following still doesn't work.

```ballerina
public function main() {
    int:Signed8 _ = -127; //  `incompatible types: expected 'int:Signed8', found 'int'``
    int:Signed8|int:Unsigned32 _ = -127; // `incompatible types: expected '(int:Signed8|int:Unsigned32)', found 'int'` and `operator '-' not defined for '127'`
}
```

**Pull request closed without merging**